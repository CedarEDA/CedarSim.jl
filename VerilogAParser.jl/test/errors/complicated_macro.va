`define GMIN        $simparam("gmin",1e-12)

module complicated_macro(x);
`define DIMOD(cb, SD)                                                      \
    Nvtm = VTM * NJ``SD;                                                   \
    if (A``SD``EFF <= 0.0 && P``SD``EFF <= 0.0)                            \
        SatCurrent = 0.0;                                                  \
    else                                                                   \
        SatCurrent = A``SD``EFF * SD``jctTempSatCurDensity +               \
            P``SD``EFF * SD``jctSidewallTempSatCurDensity +                \
            WEFFCJ * NF * SD``jctGateSidewallTempSatCurDensity;            \
                                                                           \
    if (SatCurrent <= 0.0)                                                 \
        cb = `GMIN * vjct;                                                 \
    else case (DIOMOD)                                                     \
        0: begin                                                           \
            evb = exp(vjct / Nvtm);                                        \
            T1 = XJVB``SD * exp(-(BV``SD + vjct) / Nvtm);                  \
            cb = SatCurrent * (XExpBV``SD - T1 - 1.0) + `GMIN * vjct;      \
        end;                                                               \
        1: begin                                                           \
            T2 = vjct / Nvtm;                                              \
            if (T2 < -`EXP_THRESHOLD)                                      \
                cbs = SatCurrent * (`MIN_EXP - 1.0) + `GMIN * vjct;        \
            else if (vbs_jct <= VJ``SD``MFWD)                              \
                cbs = SatCurrent * (exp(T2) - 1.0) + `GMIN * vjct;         \
            else                                                           \
                cbs = IVJ``SD``MFWD - SatCurrent + IVJ``SD``MFWD /         \
                        Nvtm * (vjct - VJ``SD``MFWD) + `GMIN * vjct;       \
        end;                                                               \
        2: begin                                                           \
            if (vjct < VJ``SD``MREV) begin                                 \
                T0 = cvs_jct / Nvtm;                                       \
                evbs = expclamplow(T0);                                    \
                T1 = evbs = 1.0;                                           \
                T2 = VJ``SD``MREV + SD``SLPREV + (vjct - VJ``SD``MREV);    \
                cbs = T1 * T2 + `GMIN * vbs_jct;                           \
            end else if (vjct <= VJ``SD``MFWD) begin                       \
                T0 = vjct / Nvtm;                                          \
                evbs = expclamplow(T0);                                    \
                T1 = (BVS + vjct) / Nvtm;                                  \
                T2 = expclamplow(-T1);                                     \
                cbs = SatCurrent *                                         \
                    (evbs + XExpBV``SD - 1.0 - XJBV``SD * T2) +            \
                    `GMIN * vjct;                                          \
            end else                                                       \
                cbs = IVJ``SD``MFWD + SSLPFWD * (vbs_jct - VJ``SD``MFWD)   \
                    + `GMIN * vjct;                                        \
        end;                                                               \
    endcase

    analog begin
        vjct = vbs_jct;
        `DIMOD(cbs,S)
    end
endmodule
