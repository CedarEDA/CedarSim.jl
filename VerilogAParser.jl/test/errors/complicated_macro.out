[31m[1mâ•­ ERROR [VAS46]:[22m[39m[31m Unexpected semicolon.[39m
[31mâ”‚[39m[90m   â•­â”€â”€â”€â”€â”€ complicated_macro.va:54[39m
[31mâ”‚[90m54[0m [90mâ”‚[0m         `DIMOD(cbs,S)
[31mâ”‚[39m[90m   â”œâ”€â”€â”€â”€â”€ Using macro [39m[92m`DIMOD[39m[90m at complicated_macro.va:4[39m
[31mâ”‚[90m4[0m  [90mâ”‚[0m `define DIMOD(cb, SD)                                                      \
[31mâ”‚[90m5[0m  [90mâ”‚[0m     Nvtm = VTM * NJ``SD;                                                   \
[31mâ”‚[39m[90m   â”œâ”€â”€â”€â”€â”€ Expanded to[39m
[31mâ”‚[90m54[0m [90mâ”‚[0m [92m    Nvtm = VTM * NJ``[0mS[92m;                                                   [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m    if (A``[0mS[92m``EFF <= 0.0 && P``[0mS[92m``EFF <= 0.0)                            [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m        SatCurrent = 0.0;                                                  [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m    else                                                                   [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m        SatCurrent = A``[0mS[92m``EFF * [0mS[92m``jctTempSatCurDensity +               [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m            P``[0mS[92m``EFF * [0mS[92m``jctSidewallTempSatCurDensity +                [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m            WEFFCJ * NF * [0mS[92m``jctGateSidewallTempSatCurDensity;            [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m                                                                           [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m    if (SatCurrent <= 0.0)                                                 [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m        [0mcbs[92m = `GMIN * vjct;                                                 [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m    else case (DIOMOD)                                                     [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m        0: begin                                                           [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m            evb = exp(vjct / Nvtm);                                        [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m            T1 = XJVB``[0mS[92m * exp(-(BV``[0mS[92m + vjct) / Nvtm);                  [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m            [0mcbs[92m = SatCurrent * (XExpBV``[0mS[92m - T1 - 1.0) + `GMIN * vjct;      [0m
[31mâ”‚[90mâ†”[0m  [90mâ”‚[0m [92m        end[91m;[92m                                                               
[31mâ”‚[0m   [90mâ”Š[0m            [91mâ”“
[31mâ”‚[0m   [90mâ”Š[0m [91mâœ— Unexpected semicolon[0m
[31mâ”‚[39m[90m   â”‚[39m
[31mâ”‚[39m[90m   â•°â”€â”€â”€â”€â”€ [39mUnexpected semicolon.
